### PR TITLE
Precalculate scores for all possible subsequences in getReactivityScore

### PR DESCRIPTION
**Pull request by fymue**
<details open>
<summary>Description of initial idea

</summary>

While thinking about how the reactivity score is actually being calculated in the `calculateScore` function, I found another way to calculate the score that is even quicker than the combination of the caching/lookups from PR #53 and the avoidance of unnecessary string comparisons from PR #54. I realized that `calculateScore` essentially just iteratively sums up the `probingData` values from the `.shape`-files (let's call them "basescores") for the bases of the current `Subsequence`. This sum can also be calculated by subtracting the sum of the basescores from the beginning of the sequence to the base at position `i` of the current `Subsequence` from the sum of the basescores from the beginning of the sequence to the base at position `j` of the current `Subsequence`:

```
# "Sequence" stands for the entire input sequence
score(Subsequence(i, j)) = basescore_sum(Sequence(0, j)) - basescore_sum(Sequence(0, i))
```

 This means that instead of calculating the score for a `Subsequence` iteratively in O(n) time, we can also calculate the same score in O(1) if we precalculate the sums of the basescores up to every base of the input sequence, which can be done in O(n). For clarification consider the following example:

 ```
 # for simplicity, let's ignore the complete calculateScore implementation
 # and assume it just sums up the values
 # (the algorithm still works the same with the real calculateScore implementation)

 Sequence = "AUGCUG"
 #               A    U     G     C    U    G
 probingData = [0.4, 0.6, -0.5, -0.1, 0.8, 0.7]

 subseq_1 = Sequence[1:5]  #  UGCU
 subseq_2 = Sequence[0:3]  #  AUG
 subseq_3 = Sequence[3:6]  #  CUG

 def calculateScore(i, j):
    # i and j are the start and end bases/positions
    # of the Subsequence (j is not inclusive)
     sum = 0
     for x in range(i, j):
         sum += probingData[x]
     return sum

 score_1 = calculateScore(1, 5)  # 0.6 - 0.5 - 0.1 + 0.8 = 0.8
 score_2 = calculateScore(0, 3)  # 0.4 + 0.6 - 0.5 = 0.5
 score_3 = calculateScore(3, 6)  # -0.1 + 0.8 + 0.7 = 1.4

# generating baseScoreSums in O(n)
 sum = 0
 baseScoreSums = [0]
 for value in probingData:
     sum += value
     baseScoreSums.append(sum)

# baseScoreSums = [0, 0.4, 1, 0.5, 0.4, 1.2, 1.9]

def fast_score(i, j):
    # i and j are the start and end bases/positions
    # of the Subsequence (j is not inclusive)
    return baseScoreSums[j] - baseScoreSums[i]

fast_score_1 = fast_score(1, 5) = 0.8  #  1.2 - 0.4 = 0.8
fast_score_2 = fast_score(0, 3) = 0.5  #  0.5 - 0 = 0.5
fast_score_3 = fast_score(3, 6) = 1.4 #  1.9 - 0.5 = 1.4

# fast_score and calculateScore return the same scores
# while fast_score can calculate them in O(1) instead of O(n)
```

I implemented this new approach to calculate the reactivity scores during the probing, compared it to all the major probing speedups from the previous PRs (#53, #54) and got the following results:

nodangle_probing_final

As you can see, this implementation is noticeably faster than the last major improvement (avoid unnecessary string comparions, #54, blue line) and about twice as fast as the first improvement (lookups/caching of previously calculated scores, #53, red line). And since we now only need to calculate and store the basescore basescore sums for every base of the input sequence (O(n)), the program also requires way less memory than the other improvements (which used an O(n²) cache), as you can see in the memory usage plot below.

nodangle_probing_final_space

I also profiled the program again to compare the percentages of the total execution time that can be attributed to the `getReactivityScore` function between this improvement and the original implementation. 

Current, improved version:

getReactivityScore_profiling_current

Original version:

getReactivityScore_profiling_original

As you can see in the partial profiling call graphs above, the `getReactivityScore` function now only accounts for 20.5% of the total execution time, compared to 93.6% when using the original implementation.
</details>

<details>
<summary>
Second idea (Now no longer necessary since we accept the minor numeric differences)

</summary>

~~Since my initial idea (see collapsed section) unfortunately didn't end up working due to the ever so slightly different score values,~~  I went ahead and adjusted the caching/lookup implementation to now fully precalculate the reactivity scores of all possible subsequences instead of adding them dynamically to the lookup array.

 Since I was able to transfer parts of the  algorithm of my initial idea, precalculating all possible scores now happens in a few milliseconds, which means that after the initial call to `getReactivityScore`, the function only performs lookups and doesn't need to do any conditional checks or index calculation like [before](https://github.com/jlab/fold-grammars/pull/54). The resulting speedup is noticeable, but unfortunately not quite as big as the speedup of my initial idea.

<details>
<summary>
Possible Explanation for the slow-down

</summary>

I assume this slow-down is causes by cache misses, which are very likely to occur here since the input subsequences (or rather their `i` and `j` positions) are somewhat random., meaning `getReactivityScore` doesn't receive continuous subsequences after a pattern like `Subsequence(0, 1)`, `Subsquence(0,2)`... As a result the next required score isn't part of the scores that were prefetched by the CPU during the last call to `getReactivityScore`, so they have to be flushed from the CPU cache and the required score has to be fetched from RAM which takes significantly longer. This wasn't a problem in my initial idea/implementation, because it only required and `n`-sized array, which can probably be entirely stored in the CPU cache and thus doesn't cause any cache misses when lookups are performed.
</details>

time

As you can see, the execution time of the updated implementation (yellow line) sits somewhat in between the execution time of my initial implementation of this PR (green line)  and the currently used implementation from PR #56 (blue line). 

space

Since I now need to precalculate/store all possible scores again, the required memory equals the required memory of the currently used implementation again.

Screenshot from 2022-12-08 01-33-51

The partial profiling call graph shows that now about 25% of the total execution time is spent in `getReactivityScore` (cf. percentages in first collapsed section). For reference, if we instantly return from `getReactivityScore`, it still accounts for about 19% of the total execution time of the program. So even though this Gapc compilation still spends a significant amount of time in this function, its impact on the total execution is almost as small as it could theoretically be.
</details>

### CODE REVIEW

**Comment by fymue** [comment]
Ah, that error happens, because I didn't consider the case that the `.shape` file contains more entries than the sequence length. If you adjust the `for` loop condition in  `calcBaseScores` function a tiny bit you will no longer get an error:

```
const unsigned int seqLen = Base.seq->n;
for (unsigned int i = 0; i < probingData.size() && i < seqLen; i++) {
...
```

This might actually also happen in the latest implementation of this PR, I will have to double check that. So thanks for pointing this out.

**Comment by sjanssen2** [comment]
I've now inspected the full folding space of the input sequences and can only spot minor numeric differences in the range of < 10^-16. Thus, I'd say https://github.com/jlab/fold-grammars/commit/11df1f39b848bdcbf65c2228d00849b994347850 is a success and we "just" have to update the Truth files for testing!

**Comment by fymue** [comment]
> I've now inspected the full folding space of the input sequences and can only spot minor numeric differences in the range of < 10^-16. Thus, I'd say [11df1f3](https://github.com/jlab/fold-grammars/commit/11df1f39b848bdcbf65c2228d00849b994347850) is a success and we "just" have to update the Truth files for testing!

Sounds great! Then I will revert back to  [11df1f3](https://github.com/jlab/fold-grammars/commit/11df1f39b848bdcbf65c2228d00849b994347850) and apply the bug fix on top.

File: Extensions/probing.hh [hunk 1]
```diff
@@ -232,94 +232,87 @@ static double CalculatePseudoEnergy(double data, const std::string &modifier,
...
+  // allocate space and store pointers in baseScores
+  baseScores[0] = new double[lookupSize];  // paired base scores
+  baseScores[1] = new double[lookupSize];  // unpaired base scores
+  baseScores[0][0] = 0.0;
+  baseScores[1][0] = 0.0;
+
+  double pairedScore = 0.0, unpairedScore = 0.0;
+
+
+  for (unsigned int i = 0; i < probingData.size() && i < seqLen; i++) {
```
Comment (sjanssen2) [inline Extensions/probing.hh:260]: will this also work for co-folding situations, where the probing data file contains a separator to split probing information for the two input nucleotide sequences?

Comment (fymue) [reply Extensions/probing.hh:260]: I didn't change they way the probing information is parsed, so this should also work in those situations.

**Comment by sjanssen2** [comment]
@kmaibach ping :-) Can you please let us know your opinion on this PR

**Review by kmaibach — APPROVED** [review]
My apologies. With all the PRs, I missed this one.

The speedup looks great and if the differences are that minor I agree that this is a great improvement. Also, this will help with my calculations as well regarding the runtime.

**Pull request merged**